pages/docs/react/latest/memo.mdx

```diff
@@ -0,0 +1,90 @@
+---
+title: memo
+description: "using React.memo"
+canonical: "/docs/react/latest/memo"
+---
+
+# memo
+
+`React.memo` lets you skip re-rendering a component when its props are unchanged.
+
+Wrap a component in memo to get a memoized version of that component.
+This memoized version of your component will usually not be re-rendered when its parent component is re-rendered as long as its props have not changed.
+
+<small>But React may still re-render it: memoization is a performance optimization, not a guarantee.</small>
+
+<CodeTab labels={["ReScript", "JS Output"]}>
+
+```res
+@react.component
+let make = React.memo((~a: int, ~b: string) => {
+  <div>
+    {React.int(a)}
+    <br />
+    {React.string(b)}
+  </div>
+})
+```
+
+```js
+import * as React from "react";
+import * as JsxRuntime from "react/jsx-runtime";
+
+var make = React.memo(function (props) {
+      return JsxRuntime.jsxs("div", {
+                  children: [
+                    props.a,
+                    JsxRuntime.jsx("br", {}),
+                    props.b
+                  ]
+                });
+    });
+```
+
+</CodeTab>
+
+## arePropsEqual
+
+In React, memo can accept an optional argument called "arePropsEqual". This function takes two arguments: the previous props and the new props of the component.
+It should return true if the old and new props are the same, meaning the component will produce the same output and behavior with the new props as it did with the old ones.
+
+In ReScript, to use the `arePropsEqual` function, you must redefine the `make` binding with `React.memoCustomCompareProps`.
```

Comment:
Not sure about "must" here, because, as you demonstrated, there are other ways as well.
Maybe "you can redefine the `make` binding"?

Reply:
There is a bit of "must" in the sense that you cannot just write:

https://rescript-lang.org/try?version=v11.1.1&code=AIJwpghgxgLgdFA9gWwA6IHZgzAUAGzBgAJkIBrMYgXmICVJY5kxlEBhAVwGcYV2UqCOAAKIRKm4AKXMWJSAfgBMAltwgAjQkoA0xBZnb4VUcgEoaAPmIBvWXOIAeDZxh8M9h8VXqtYJdQ2PpraAL6eDobGpoFgAG5WxPEAtJYAUtwAHgCicdjwALKIPGBwqOB5OAAiYABmEJz4MKlRJuShlhFyNgzQ8LwgKhgA5lIARAUAnsQubphjZuFejgD0s+6dcqE69lJSqACMeqgATBbU1odwwX5KNLSn12oh-ma4b0A

So, you must do something other than that to get it to work. Which is the exact thing I want to add to the documentation, as that was confusing to me.

Reply:
The other thing you can do is to not use the JSX preprocessor extension at all btw (i.e. no `@react.component` annotation). But that once again requires explicitly typing props.

```rescript
type props = {
  disabled: bool,
  onClick: JsxEvent.Mouse.t => unit,
}

let make = React.memoCustomCompareProps(
  ({disabled, onClick}) => {
    <button
      disabled={disabled}
      onClick={ev => ev->JsxEvent.Mouse.preventDefault->onClick}>
      {React.string("My button")}
    </button>
  },
  (p1, p2) => p1.disabled == p2.disabled,
)
```

so maybe we can keep the "must", but give an alternative solution

Reply:
I rephrased it to emphasize you can't combine `@react.component` and `React.memoCustomCompareProps`. Added the two workarounds as example.

Reply:
For my own understanding, does `@react.component` do anything in terms of using the component in JSX syntax? Or is that not a requirement?

Reply:
Since JSX4, what the annotation does is basically generating a props type (probably with more generic types though). So no, using them is identical.
Have a look at the [blogpost that announced JSX V4 in ReScript 10.1](https://rescript-lang.org/blog/release-10-1#new-jsx-v4-syntax).

Reply:
I see, thank you!